[DI] Fix bad error message for unused bind under _defaults

przemyslaw-bogusz opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #27828 
| License       | MIT

**Sidenote**: I originally included the fix in #29897, but I decided to close the previous PR and divide it into two separate PRs for clarity.

**Description:**
With this fix, the message regarding an unused bind will have a clear information about the type of the bind (defined under __defaults_, __instanceof_ or _per service_), as well as the name of the file, in which it was configurated. It's for, both, YAML and XML configurations.

For the core team, please note, that the fix assumes a possibility of definings binds under __instanceof_, which was introduced in #27806. But since fixes are merged into other branches, I thought that it might be necessary to inlude this possibility. If this case requires making separate fixes for different branches, I will gladly do it.

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveBindingsPass.php:
```diff
@@ -37,8 +37,21 @@ public function process(ContainerBuilder $container)
-            foreach ($this->unusedBindings as list($key, $serviceId)) {
-                $message = sprintf('Unused binding "%s" in service "%s".', $key, $serviceId);
+            foreach ($this->unusedBindings as list($key, $serviceId, $type, $file)) {
+                $message = sprintf('You have a "bind" configured for an argument named "%s" ', $key);
```
`for an argument named "%s"`: that part is misleading: binding can also work by type. - and also by type+$name since 4.2. We should adjust the message for that.

przemyslaw-bogusz replied:
Done

nicolas-grekas reviewed (commented):
Looks nice!
That should target master to me, it's more an improvement than a bug fix.
Tests fail because composer.json files need to bump the minimum supported version of di, for http-kernel.

sstok commented:
Can this also be added for the PHP DSL loader? Really nice addition!

przemyslaw-bogusz committed:
[DI] Fix bad error message for unused bind under _defaults

przemyslaw-bogusz commented:
Thanks for the positive feedback!

@sstok 
I added the same functionality to PHP DSL loader.

@nicolas-grekas 

1. Changed target to master.
1. Modified the error message so that it refers to argument's name, type or both - depending on the configuration.

przemyslaw-bogusz commented:
@nicolas-grekas 
> Tests fail because composer.json files need to bump the minimum supported version of di, for http-kernel.

Does it mean I have to make changes in composer.json in my repository? It seems to be identical with the one from Symfony's repository?

fabpot commented:
@nicolas-grekas friendly ping

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Argument/BoundArgument.php:
```diff
@@ -16,35 +16,48 @@
+    const DEFAULTS_BIND = 1;
+    const INSTANCEOF_BIND = 2;
+
+    public static $messages = [
```
I'd suggest removing this variable, it's not needed.

przemyslaw-bogusz replied:
Done

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Argument/BoundArgument.php:
```diff
@@ -16,35 +16,48 @@
 {
+    const SERVICE_BIND = 0;
+    const DEFAULTS_BIND = 1;
+    const INSTANCEOF_BIND = 2;
```
BINDING instead of BIND?

przemyslaw-bogusz replied:
Done

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Compiler/ResolveBindingsPass.php:
```diff
@@ -37,8 +37,39 @@ public function process(ContainerBuilder $container)
+                    $message .= sprintf(' in file "%s"', $file);
+                }
+
+                $message .= '. But, this argument was not found in any of the services it was applied to. It may be unused and can be removed, or it may have a typo.';
```
When $serviceId is used, this part could be improved as there is only one.
I'd also suggest to make the first part a single sentence (... but ...) e.g:
`A binding is configured for arguments [...] but no corresponding arguments have been found. It may be unused and should be removed, or it may have a typo.`

przemyslaw-bogusz replied:
The exception is thrown inside the foreach loop, so it actually contains information about a single binding. Should I change it to include all? What if there is a large amount of them?

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/Configurator/ServicesConfigurator.php:
```diff
@@ -137,4 +139,9 @@ public function __construct(ContainerBuilder $container, PhpFileLoader $loader,
         return $this->set($id, $class);
     }
+
+    final public function getPath()
```
can we get rid of the method? It'd be better to find another way to communicate the path to the outside and not pollute autocompletion (since this is what this configurator is about). Maybe passing as constructor arguments?

przemyslaw-bogusz replied:
Hope this is what you had in mind

nicolas-grekas commented on src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php:
```diff
@@ -532,6 +534,13 @@ private function parseDefinition($id, $service, $file, array $defaults)
+                $bindingType = $this->isLoadingInstanceof ? BoundArgument::INSTANCEOF_BIND : BoundArgument::SERVICE_BIND;
+                foreach ($bindings as $argument => $value) {
+                    if ($value instanceof BoundArgument) {
+                        continue;
```
no need for the continue here, let's move the next line inside the "if"

przemyslaw-bogusz replied:
Done

nicolas-grekas commented on src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php:
```diff
@@ -138,7 +139,7 @@ public function process(ContainerBuilder $container)
 
                         list($bindingValue, $bindingId) = $binding->getValues();
-                        $binding->setValues([$bindingValue, $bindingId, true]);
+                        $binding->setValues([$bindingValue, $bindingId, true, BoundArgument::SERVICE_BIND, null]);
```
why not keep the type and file returned the line just before?

przemyslaw-bogusz replied:
Done

nicolas-grekas reviewed (commented):
Looks nice, sorry for the late review, here are some comments.

weaverryan commented:
Ping @przemyslaw-bogusz! Will you have time to check out the tweaks? We can hopefully get this into 4.3! :)

przemyslaw-bogusz commented:
I will do it tomorrow, so I believe we can close it by the end of the weekend. Hope that's good enough.

weaverryan commented:
It's great :). Thank you for your work!

Simperfit commented:
@przemyslaw-bogusz do you have time to finish this or do you want me to take the rest of it ?

przemyslaw-bogusz commented:
@nicolas-grekas, @weaverryan Thank you for your patience. If you have any additional comments, please let me know. I hope we will also be able to close #29944.

nicolas-grekas commented:
Thank you @przemyslaw-bogusz.

nicolas-grekas commented:
Thank you @przemyslaw-bogusz.

weaverryan commented:
Congrats! Thank you SO much @przemyslaw-bogusz!!

scott-r-lindsey commented:
Workaround until 4.2.6: make sure that at least one service consumes each dependency mentioned in _defaults / bind

weaverryan commented:
That will still be the case after this PR. This is just to correct the error message, which was misleading. I think #29944 might be what you’re looking for.

przemyslaw-bogusz commented:
@scott-r-lindsey I am not sure, if I understand you correctly, but the whole purpose of `ResolveBindingsPass` is to inform you of any unused bindings - the ones that you define, but do not use/consume in any of your services. You simply cannot have a surplus of bindings.
